### PR TITLE
Fix #9

### LINKED ISSUE #9: Fix movie preview image generation for movies of 4 seconds or shorter
Preview images for movies, used when browsing the movie library in motionEye's UI, fail to generate for movies with less than 4 seconds length (using default settings):
- They are generated at an offset of 2x the selected number of frames before the motion even that shall be included: "Kaptita Antaŭe"/"Captured Before" setting, but <s>max</s> minimum at 4 seconds.
- The default value for those "Captured Before" frames is 1, and the default framerate is 2, which leads to a default offset for preview images of 4 seconds.
- If the video has less than 4 seconds length, FFmpeg throws an error, and the video won't have any (functional) preview image, causing further server errors when browsing the library.
- Older versions of FFmpeg seem to not have generated an error, but a 0 bytes image file instead. motionEye handled this case, and retried image generated at an offset of 0. Newer FFmpeg however trows an error (non-zero exit code), in which case motionEye aborts the attempt.

There are two ways to solve the issue:
- Retry at 0 bytes offset as well in case of `ffmpeg` command call error, or in case further narrowed down condition to exclude errors caused by other reasons.
- Check movie length and use a valid offset in the first place. This is surely the cleaner solution, but requires a way to obtain the movie length. Since `ffmpeg` is used, `ffprobe` is almost assured to be available, which can be used for this purpose. A Python library would be an alternative, but additional non-std dependencies should be probably avoided.

Relevant code section: https://github.com/swe-productivity/motioneye/blob/main/motioneye/mediafiles.py#L329

Upstream issue: https://github.com/motioneye-project/motioneye/issues/3141

### PR DESCRIPTION
**Pull request by strawberry-poodle**
Fixes #9 by checking video duration using ffprobe before capturing a thumbnail.

### CODE REVIEW

File: motioneye/mediafiles.py [hunk 1]
```diff
@@ -325,12 +326,31 @@ def cleanup_media(media_type: str) -> None:
         )
         _remove_older_files(target_dir, preserve_moment, clean_cloud_info, exts=exts)
 
+def get_movie_duration_seconds(path):
+    cmd = f'ffprobe -v error -show_entries format=duration -of default=noprint_wrappers=1:nokey=1 {path}'
```
Comment (MichaIng) [inline motioneye/mediafiles.py]: It might be an idea to pass `path` as separate argument, to assure it is not split if it contains spaces. Or `cmd` could be defined as list in the first place, skipping `cmd.split()`.

File: motioneye/mediafiles.py [hunk 2]
```diff
@@ -325,12 +326,31 @@ def cleanup_media(media_type: str) -> None:
...
+    
+    try:
+        result = utils.call_subprocess(cmd.split(), stderr=subprocess.STDOUT)
```
Comment (MichaIng) [inline motioneye/mediafiles.py]: Is `stderr=subprocess.STDOUT` reasonable or somehow needed here? We want to have STDOUT only in `result`, and STDERR could either be muted `stderr=DEV_NULL` (which is the default), or shown at `stderr=subprocess.STDERR`. Currently not sure whether this lands on the console or system journal respectively, and would double with the content of `CalledProcessError` or not.

File: motioneye/mediafiles.py [hunk 3]
```diff
@@ -325,12 +326,31 @@ def cleanup_media(media_type: str) -> None:
...
+        duration_seconds = math.floor(float(result))
```
Comment (MichaIng) [inline motioneye/mediafiles.py]: Since the value is expected to be positive, no need to use `math.floor()`:
```suggestion
        duration_seconds = int(float(result))
```

File: motioneye/mediafiles.py [hunk 4]
```diff
@@ -23,6 +23,7 @@
 import re
 import subprocess
 import typing
+import math
```
Comment (MichaIng) [inline motioneye/mediafiles.py]: Related to above comment, this could then be removed again.
```suggestion
```

File: motioneye/mediafiles.py [hunk 5]
```diff
@@ -325,12 +326,31 @@ def cleanup_media(media_type: str) -> None:
...
     framerate = camera_config['framerate']
     pre_capture = camera_config['pre_capture']
     offs = pre_capture / framerate
-    offs = max(4, offs * 2)
+    movie_duration = get_movie_duration_seconds(full_path)
+    
+    if(movie_duration >= 4):  
+        offs = max(4, offs * 2)
+    else:
+        offs = offs * 2
```
Comment (MichaIng) [inline motioneye/mediafiles.py]: _I just recognized I mixed up `max` with `min` in the issue description. So it uses an offset of 2*pre-capture duration, but at least (not at most) 4 seconds._

Anyway, the default pre-capture duration is 2 seconds, hence the change you applied here would not change something with default settings, but only if the number of pre-capture frames is reduced or framerate is increased.

Thinking about it, any movie recorded by `motion` in case of a motion event, has at least the pre-capture duration that was configured at the time of the recording. 2*offs is intended to show a preview image at pre-capture duration _after_ the motion event. Whether the motion event has that much frames is however not guaranteed. While recordings would usually have at least the pre-capture duration, even if the motion event was a single frame only, settings could have changed in the meantime, and it is of course possible to just move other movie files into the camera data directory. Means, we cannot rely on any movie length. The minimum of 4 seconds seems to be arbitrary as well.

I would suggest a staged approach:
- If the recording is >=2*pre-capture duration, keep it as before
- If it is instead just >=1*pre-capture duration, use that instead, which would usually be the first frame at the start of the motion event.
- Else, use actual duration/2, hence a frame from the mid of the video.

Can be done in a single command:

```suggestion
    offs = max(x for x in {offs*2, offs} if x <= movie_duration, default=movie_duration/2)
```

**Review by MichaIng — CHANGES_REQUESTED** [review]

**Comment by strawberry-poodle** [comment]
Thanks for all the help @MichaIng! I implemented all your suggestions plus extended the exception handling for the ffprobe call. On some video containers or exceptionally short files (I somehow managed to make one with a single frame?) it can respond with N/A that breaks the float() function at mediafiles.py:333

**Comment by MichaIng** [comment]
@strawberry-poodle 
Great, also thanks for testing the edge case with a single frame.

Just one more question: Was there a particular reason you use `stderr=subprocess.STDOUT` for the `ffprobe` call? Otherwise I would suggest to either use `stderr=subprocess.STDERR`, or, if the exception doubles STDERR, remove the argument (which means STDERR is muted).

**Comment by strawberry-poodle** [comment]
Good question! If I recall correctly, the idea was to make sure we actually capture any exception strings if ffprobe itself fails. I just did a little extra testing, and indeed without `stderr=subprocess.STDOUT`, in the exception, `e.output` is empty in `subprocess.CalledProcessError`. Something that could potentially cause ffprobe to fail is if the file we're trying to generate a thumbnail for is not found. But in that case, the [ffmpeg call will already capture STDERR and reports it in the exception](https://github.com/strawberry-poodle/motioneye/blob/c9364d689b82eb455a9721ff88101f00467f163b/motioneye/mediafiles.py#L368), so maybe there's no point having the same error logged twice. So whether to keep `stderr=subprocess.STDOUT` depends on whether we're okay with potential double log entries for each failed thumbnail generation.

**Comment by MichaIng** [comment]
Okay I played around with it. When using `stderr=subprocess.STDOUT`, and the command succeeds but prints something to STDERR, it messes the actual length it returns. Sometimes there might be warnings printed to STDERR while STDOUT contains the correct length. So `stderr=subprocess.STDOUT` should not be used if the output is processed afterwards the way we do. `subprocess.STDERR` does not exist, weirdly, currently not sure how to print STDERR of the process to STDERR of Python that way. So for now, the only reasonable way is to omit the `stderr` argument and leave STDERR of `ffprobe` muted. We can find a way to preserve STDERR to be attached to the exception text or printed to logs another time.

There is also a code error now that the command is list: `.split()` is not needed anymore but fails, of course.

EDIT: Actually `stderr=subprocess.PIPE` would preserve STDERR. `subprocess.STDOUT` indeed means that STDERR should go to the same target as STDOUT, similar to `2>&1` in shell. So `subprocess.STDERR` expectedly does not exist.

And the concern about doubled logs applies only when using `stderr=None` in which case it goes to STDERR of the Python process itself. In contrast `stderr=subprocess.PIPE` makes Python capture it, and available in `CompletedProcess.stderr` and `CalledProcessError.stderr`, which is exactly what we want.

EDIT2: Nah, in case of a `CalledProcessError`, its `stderr` is not automatically attached. This would either need to be done handling that exception specifically, or just using `stderr=None` (leaving it in console/syslog output).

File: motioneye/mediafiles.py [hunk 6]
```diff
@@ -325,12 +325,28 @@ def cleanup_media(media_type: str) -> None:
         )
         _remove_older_files(target_dir, preserve_moment, clean_cloud_info, exts=exts)
 
+def get_movie_duration_seconds(path):
+    cmd = ['ffprobe', '-v', 'error', '-show_entries', 'format=duration', '-of', 'default=noprint_wrappers=1:nokey=1', path]
+    
+    try:
+        result = utils.call_subprocess(cmd.split(), stderr=subprocess.STDOUT)
```
Comment (MichaIng) [inline motioneye/mediafiles.py]:
```suggestion
        result = utils.call_subprocess(cmd, stderr=None)
```

**Comment by strawberry-poodle** [comment]
Thank you for having another look! I applied your suggestion, I agree that `stderr=None` will work just fine. 

In the future, could probably also consider updating other subprocess calls just for consistency, I see that `stderr=subprocess.STDOUT` is used elsewhere, but I don't quite see reason why we'd redirect stderr in some cases but not in others. But also maybe good idea not to touch those other calls since they're not broken. 🙂

**Review by MichaIng — APPROVED** [review]
Great, LGTM!

Yeah the other calls should be aligned indeed. IMO most importantly `shell=True` should be removed consequently, and any processing that is probably done in shell with `grep`/`sed` or such, should be done in Python. Makes sense to also review `stderr=` values, but there can be reasons to merge STDOUT and STDERR. I remember I saw in one case that the value is validated, like whether it is an integer or so, in case taken, otherwise printed in an error message. Can be simpler, if invalid values can be in STDOUT as well as STDERR. Could have been done as well here that way, so we would have seen the `N/A` value. But the exception handles that and `{e}` at least should tell you that it was no numerical value.
